src/MessageSystem.hpp

```diff
@@ -0,0 +1,176 @@
+/*
+ *  Copyright (C) 2018 SPMod Development Team
+ *
+ *  This file is part of SPMod.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "spmod.hpp"
+
+class Message : public IMessage
+{
+public:
+    struct MessageParam
+    {
+        MsgParamType type;
+        std::variant<int, float, std::string> data;
+    };
```

Comment:
Awww, that's something weird, tagged tagged enum. Not sure how we can do it better though.